File: frontend/types/schemas/unit/Unit.ts
import { z } from "zod";

export const UnitSchema = z
  .object({
    id: z.number().int().default(0),
    name: z.string().default("Object"),
    transform: z
      .object({
        id: z.number().int().default(0),
        name: z.string().default("Object"),
        position: z.object({
          x: z.number().int().default(0),
          y: z.number().int().default(0),
        }),
        size: z.object({ height: z.number().int(), width: z.number().int() }),
      })
      .describe("An entity component can have spatial data."),
    kind: z
      .enum(["Unit", "Sprite", "Location", "Tile", "Mask"])
      .default("Unit"),
    serial_number: z.union([z.number().int(), z.null()]).default(null),
    use_default: z.boolean().default(true),
    unit_definition: z
      .object({
        id: z.number().int().default(0),
        name: z.string().default("Definition"),
        ref_type: z.literal("Definition").default("Definition"),
        use_default: z.boolean().default(true),
        specification: z.object({
          name: z.string().default("Unit Specification"),
          graphics: z.number().int(),
          subunit1: z.number().int(),
          subunit2: z.number().int(),
          infestation: z.union([z.number().int(), z.null()]),
          construction_animation: z.number().int(),
          unit_direction: z.number().int(),
          portrait: z.number().int(),
          label: z.number().int(),
        }),
        stats: z.object({
          name: z.string().default("Unit Status"),
          hit_points: z.object({
            current: z.number().int().gte(0).default(0),
            max: z.number().int().gte(0).default(0),
          }),
          shield_enable: z.boolean(),
          shield_points: z.object({
            current: z.number().int().gte(0).default(0),
            max: z.number().int().gte(0).default(0),
          }),
          energy_points: z.object({
            current: z.number().int().gte(0).default(0),
            max: z.number().int().gte(0).default(0),
          }),
          armor_points: z.number().int().lt(256).default(0),
          armor_upgrade: z.number().int(),
          rank: z.number().int(),
          elevation_level: z.number().int(),
        }),
        weapons: z.object({
          ground_weapon: z.union([
            z.object({
              id: z.number().int().default(0),
              name: z.string().default("Definition"),
              ref_type: z.literal("Definition").default("Definition"),
              damage: z.object({
                amount: z.number().int().gte(0).lte(65536),
                bonus: z.number().int().gte(0).lte(65536),
                factor: z.number().int(),
              }),
              bullet: z.object({
                behaviour: z.number().int(),
                remove_after: z.number().int(),
                attack_angle: z.number().int(),
                launch_spin: z.number().int(),
                x_offset: z.number().int(),
                y_offset: z.number().int(),
              }),
              splash: z.object({
                inner: z.number().int(),
                medium: z.number().int(),
                outer: z.number().int(),
              }),
              cooldown: z.number().int(),
              upgrade: z.number().int(),
              weapon_type: z.number().int(),
              explosion_type: z.number().int(),
              target_flags: z.number().int(),
              error_message: z.number().int(),
              icon: z.number().int(),
              graphics: z.number().int(),
            }),
            z.null(),
          ]),
          max_ground_hits: z.number().int(),
          air_weapon: z.union([
            z.object({
              id: z.number().int().default(0),
              name: z.string().default("Definition"),
              ref_type: z.literal("Definition").default("Definition"),
              damage: z.object({
                amount: z.number().int().gte(0).lte(65536),
                bonus: z.number().int().gte(0).lte(65536),
                factor: z.number().int(),
              }),
              bullet: z.object({
                behaviour: z.number().int(),
                remove_after: z.number().int(),
                attack_angle: z.number().int(),
                launch_spin: z.number().int(),
                x_offset: z.number().int(),
                y_offset: z.number().int(),
              }),
              splash: z.object({
                inner: z.number().int(),
                medium: z.number().int(),
                outer: z.number().int(),
              }),
              cooldown: z.number().int(),
              upgrade: z.number().int(),
              weapon_type: z.number().int(),
              explosion_type: z.number().int(),
              target_flags: z.number().int(),
              error_message: z.number().int(),
              icon: z.number().int(),
              graphics: z.number().int(),
            }),
            z.null(),
          ]),
          max_air_hits: z.number().int(),
          target_acquisition_range: z.number().int(),
          sight_range: z.number().int(),
          special_ability_flags: z.number().int(),
        }),
        ai: z.object({
          name: z.string().default("Unit AI"),
          computer_idle: z.number().int(),
          human_idle: z.number().int(),
          return_to_idle: z.number().int(),
          attack_unit: z.number().int(),
          attack_and_move: z.number().int(),
          internal: z.number().int(),
          right_click: z.number().int(),
        }),
        sound: z.object({
          name: z.string().default("Unit Sound"),
          ready: z.union([z.number().int(), z.null()]),
          what_start: z.number().int(),
          what_end: z.number().int(),
          piss_start: z.union([z.number().int(), z.null()]),
          piss_end: z.union([z.number().int(), z.null()]),
          yes_start: z.union([z.number().int(), z.null()]),
          yes_end: z.union([z.number().int(), z.null()]),
        }),
        size: z.object({
          name: z.string().default("Unit Size"),
          size_type: z.number().int(),
          placement_box_size: z.object({
            height: z.number().int(),
            width: z.number().int(),
          }),
          bounds: z.object({
            left: z.number().int(),
            top: z.number().int(),
            right: z.number().int(),
            bottom: z.number().int(),
          }),
          addon_position: z.union([
            z.object({
              x: z.number().int().default(0),
              y: z.number().int().default(0),
            }),
            z.null(),
          ]),
        }),
        cost: z.object({
          name: z.string().default("Unit Cost"),
          cost: z.object({
            mineral: z.number().int().gte(0).default(0),
            gas: z.number().int().gte(0).default(0),
            time: z.number().int().gte(0).default(0),
          }),
          build_score: z.number().int(),
          destroy_score: z.number().int(),
          is_broodwar: z.boolean(),
          supply: z.object({
            required: z.number().int(),
            provided: z.number().int(),
          }),
          space: z.object({
            required: z.number().int(),
            provided: z.number().int(),
          }),
        }),
      })
      .describe("Definition of unit specification."),
    owner: z
      .object({
        id: z.number().int().default(0),
        name: z.string().default("Object"),
        color: z.number().int(),
        rgb_color: z.array(z.any()).min(3).max(3),
        player_type: z.enum([
          "Inactive",
          "Computer (game)",
          "Occupied By Human Player",
          "Rescue",
          "Unused",
          "Computer",
          "Human (Open Slot)",
          "Neutral",
          "Closed Slot",
        ]),
        race: z.enum([
          "Zerg",
          "Terran",
          "Protoss",
          "Invalid (Independant)",
          "Invalid (Neutral)",
          "User Selectable",
          "Random",
          "Inactive",
        ]),
        force: z.number().int().gte(0).lt(4).default(0),
      })
      .optional(),
    resource_amount: z.number().int().default(0),
    hangar: z.number().int().default(0),
    unit_state: z.number().int().default(0),
    relation_type: z.number().int().default(0),
    related_unit: z.number().int().default(0),
    special_properties: z.number().int().default(0),
    valid_properties: z.number().int().default(0),
  })
  .describe(
    "Unit placed on map.\n\nThe entity means what placeable on map, so every `Unit` which herit `Entity` is placed unit.\nIf you looking for specificaiton of unit like `Max HP`, `Size`, see `UnitDefinition`.",
  );
export type Unit = z.infer<typeof UnitSchema>;

File: backend/app/models/unit.py
from typing import Optional
from app.models.definitions.unit_definition import UnitDefinition
from app.models.structs.stat import Stat
from pydantic import Field
from .player import Player
from .entity import Entity
from .structs.cost import Cost
from .structs.spatial import Position2D
from .wobject import WObject
from .components.transform import TransformComponent
from enum import Flag


class CHKUnit(Entity):
  """CHK compatible Unit."""

  serial_number: Optional[int] = None
  """Identical number when unit placed on map. -1 When non-placed unit."""
  cost: Cost
  hit_points: Stat = Stat()
  shield_points: Stat = Stat()
  energy_points: Stat = Stat()
  armor_points: int = Field(default=0, lt=256)
  owner: Player = Player(
    player_type="Inactive", race="Inactive", color=0, rgb_color=(0, 0, 0)
  )
  resource_amount: int = 0
  hangar: int = 0
  unit_state: int = 0
  relation_type: int = 0
  related_unit: int = 0
  special_properties: int = 0
  valid_properties: int = 0
  use_default: bool = True
  # transform = TransformComponent(position=Position2D(x=0, y=0))


class Unit(Entity):
  """Unit placed on map.

  The entity means what placeable on map, so every `Unit` which herit `Entity` is placed unit.
  If you looking for specificaiton of unit like `Max HP`, `Size`, see `UnitDefinition`.
  """

  serial_number: Optional[int] = None
  """Identical number when unit placed on map. -1 When non-placed unit."""
  use_default: bool = True
  unit_definition: UnitDefinition

  owner: Player = Player(
    player_type="Inactive", race="Inactive", color=0, rgb_color=(0, 0, 0)
  )
  resource_amount: int = 0
  hangar: int = 0
  unit_state: int = 0
  relation_type: int = 0
  related_unit: int = 0
  special_properties: int = 0
  valid_properties: int = 0


class PlacedUnitRelationFlag(Flag):
  nydus_link = 0b10000000
  addon_link = 0b100000000


class SpecialPropertiesFlag(Flag):
  cloak = 0b1
  burrow = 0b10
  in_transit = 0b100
  hallucinated = 0b1000
  invincible = 0b10000


class ValidPropertiesFlag(Flag):
  owner_player = 0b1
  hp = 0b10
  shields = 0b100
  energy = 0b1000
  resource = 0b10000
  amount = 0b100000


class UnitStateFlag(Flag):
  cloaked = 0b1
  burrowed = 0b10
  is_transit = 0b100
  hallucinated = 0b1000
  invincible = 0b10000


class UnitProperty(WObject):
  """Create units with properties trigger used."""

  special_properties: int
  unit_data: int
  owner: int = Field(default=0, le=1)
  hit_point_percent: int = Field(default=1, le=100, ge=0)
  shield_point_percent: int = Field(default=1, le=100, ge=0)
  energy_point_percent: int = Field(default=1, le=100, ge=0)
  resource_amount: int
  units_in_hangar: int
  flags: int


class UnitRestriction(WObject):
  availability: list[bool]
  global_availability: bool
  uses_defaults: list[bool]
